subsys/bluetooth/nrf_bt_adv/core.c

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <bluetooth/nrf_bt_adv.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(nrf_bt_adv, CONFIG_NRF_BT_ADV_LOG_LEVEL);
+
+enum provider_set {
+	PROVIDER_SET_AD,
+	PROVIDER_SET_SD
+};
+
+static void get_section_ptrs(enum provider_set set,
+			     const struct nrf_bt_adv_data_provider **start,
+			     const struct nrf_bt_adv_data_provider **end)
+{
+	switch (set) {
+	case PROVIDER_SET_AD:
+		{
+			extern const struct nrf_bt_adv_data_provider _nrf_bt_adv_ad_list_start[];
+			extern const struct nrf_bt_adv_data_provider _nrf_bt_adv_ad_list_end[];
+
+			*start = _nrf_bt_adv_ad_list_start;
+			*end = _nrf_bt_adv_ad_list_end;
+		}
+		break;
+
+	case PROVIDER_SET_SD:
+		{
+			extern const struct nrf_bt_adv_data_provider _nrf_bt_adv_sd_list_start[];
+			extern const struct nrf_bt_adv_data_provider _nrf_bt_adv_sd_list_end[];
+
+			*start = _nrf_bt_adv_sd_list_start;
+			*end = _nrf_bt_adv_sd_list_end;
+		}
+		break;
+
+	default:
+		/* Should not happen. */
+		__ASSERT_NO_MSG(false);
+		*start = NULL;
+		*end = NULL;
+		break;
+	}
+}
+
+static size_t get_section_size(enum provider_set set)
+{
+	const struct nrf_bt_adv_data_provider *start;
+	const struct nrf_bt_adv_data_provider *end;
+
+	get_section_ptrs(set, &start, &end);
+
+	return end - start;
+}
+
+size_t nrf_bt_adv_get_ad_prov_cnt(void)
+{
+	return get_section_size(PROVIDER_SET_AD);
+}
+
+size_t nrf_bt_adv_get_sd_prov_cnt(void)
+{
+	return get_section_size(PROVIDER_SET_SD);
+}
+
+static void update_common_feedback(struct nrf_bt_adv_feedback *common_fb,
+				   const struct nrf_bt_adv_feedback *fb)
+{
+	common_fb->grace_period_s = MAX(common_fb->grace_period_s, fb->grace_period_s);
```

Comment:
I don't like this grace period configuration in the core part of this module. It partially destroys the module abstraction by moving the specific configuration of the Swift Pair provider into the core layer. I think the grace period should be handled by the application code. It could be implemented in the following manner:

1. Check for the existence of the Swift Pair provider (requires a new nrf_bt_adv API that can be used to check for the presence of a specific provider using its ID number)
2. Use the grace period from the Swift Pair Kconfig when managing advertising in the application code.

Reply:
Grace period was meant to be more generic feature. It would just lead to advertising with modified payload for some additional time right before advertising is stopped. It could be also useful e.g. for Fast Pair payload (or other vendor-specific pairings). The grace period could also be used to inform remote peers that device is going to sleep.

Reply:
I think the Fast Pair does not mention any grace period mechanics. Currently, the Swift Pair is the only protocol known to me that uses it.   

Consider adding provider-specific header (e.g. ```nrf_bt_adv_swift_pair.h```) which will expose additional API features (like enabling/disabling grace period). An alternative approach would be to decide in the runtime which providers should be enabled or disabled for each run of the ```nrf_bt_adv_get_ad``` API

Reply:
I think that the grace period itself could be considered as a generic feature (advertising modified payload for predefined time before stopping advertising). The feature is optional, so may not be used if not needed.

Fast Pair does not mention grace period mechanics in docs, but I think the use-case here is quite similar, so we could use similar mechanism to prevent device from instantly "disappearing" to improve user experience (we may discuss this with Google as proposed improvement at some point).

I would prefer to avoid providing provider-specific headers, because that would significantly complicate integration and API. I would also want to let provider to decide on its own if it wants to provide data or not at given point (I prefer decentralized approach here).